src/Simulation/qdk_sim_rs/src/linalg/expm.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::{
+    fmt::Error,
+    ops::{Mul, Neg},
+    process::Output,
+};
+
+use cauchy::Scalar;
+use ndarray::{
+    linalg::Dot, Array2, ArrayBase, Data, Ix2, OwnedRepr, RawArrayView, RawData, ScalarOperand,
+};
+
+use crate::{
+    error::QdkSimError,
+    linalg::{decompositions::EigenvalueDecomposition, Identity, Inv, MatrixPower},
+    math::approximate_factorial,
+};
+
+use super::{decompositions::ExplicitEigenvalueDecomposition, HasDagger};
+
+// NB: We use the notation of
+//     https://www.cs.jhu.edu/~misha/ReadingSeminar/Papers/Moler03.pdf.
+//
+//     In particular:
+//
+//         e^{A} ≈ R_pq(A) ≔ [D_pq(A)]⁻¹ N_pq(A)
+//
+//     where
+//
+//         N_pq(A) ≔ Σ_j=0^p [(p + q - j)! p!] / [(p + q)! j! (p - j)!] A^j
+//
+//     and
+//
+//         D_pq(A) ≔ Σ_j=0^q [(p + q - j)! p!] / [(p + q)! j! (q - j)!] (-A)^j.
+
+/// Types that support the matrix exponential $e^{A}$.
+pub trait Expm<A>
+where
+    A: Scalar,
+{
+    /// Errors that can result from the [`expm`] method.
+    ///
+    /// [`expm`]: Expm::expm
+    type Error;
+    /// Type used to represent $e^{A}$ when calling [`expm`].
+    ///
+    /// [`expm`]: Expm::expm
+    type Output;
+
+    /// Returns the matrix exponential $e^{A}$ of a matrix $A$.
+    fn expm(&self) -> Result<Self::Output, Self::Error>;
+
+    /// Returns the matrix exponential $e^{\alpha A}$ of a matrix $A$,
+    /// and where $\alpha$ is a complex scalar.
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error>;
+}
+
+impl<A> Expm<A> for ExplicitEigenvalueDecomposition<A>
+where
+    A: Scalar + Mul<Output = A>,
+    ExplicitEigenvalueDecomposition<A>: EigenvalueDecomposition<A>,
+{
+    type Error = <Self as EigenvalueDecomposition<A>>::Error;
+    type Output = <Self as EigenvalueDecomposition<A>>::MatrixSolution;
+
+    fn expm(&self) -> Result<Self::Output, Self::Error> {
+        self.apply_mtx_fn(|x| x.exp())
+    }
+
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error> {
+        self.apply_mtx_fn(|x| (*x * scale).exp())
+    }
+}
+
+#[cfg(feature = "pade")]
+impl<A, S, E> Expm<A> for ArrayBase<S, Ix2>
+where
+    S: Data + RawData<Elem = A>,
+    A: Scalar + ScalarOperand,
+    E: std::convert::From<QdkSimError>,
+    ArrayBase<S, Ix2>: Inv<Error = E> + Identity,
+{
+    type Error = E;
+    type Output = ArrayBase<OwnedRepr<A>, Ix2>;
+
+    fn expm(&self) -> Result<Self::Output, Self::Error> {
+        // TODO: allow generalizing p, q
+        let p = 16;
+        let q = 16;
+
+        let n = pade_n::<A, S, E>(self, p, q)?;
+        let d_inv = pade_d::<A, S, E>(self, p, q)?.inv()?;
+
+        Ok(d_inv.dot(&n))
+    }
+
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error> {
+        Ok((self * scale).expm()?)
+    }
+}
+
+// TODO: Allow simpler type constraints.
```

Comment:
(Nit) A few months ago I saw a coding style item that I liked. It was prescribing to complement every `TODO` with, in parentheses, either user alias (who opened the `TODO` or who the `TODO` is assigned to), or the bug number, or both. E.g. `TODO(cgranade): ..`.